% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/AllGenerics.R, R/detect_TSS.R
\docType{methods}
\name{detectTSS}
\alias{detectTSS}
\alias{detectTSS,CapSet-method}
\title{Detection of Trancription start sites based on local enrichment}
\usage{
detectTSS(CSobject, groups, outfile_prefix = NULL, foldChange = 2,
  restrictChr = NULL)

\S4method{detectTSS}{CapSet}(CSobject, groups, outfile_prefix = NULL,
  foldChange = 2, restrictChr = NULL)
}
\arguments{
\item{CSobject}{CapSet object created using \code{\link{newCapSet}} function}

\item{groups}{a character vector that contains group name of the sample, for replicate-based TSS
calling (see example)}

\item{outfile_prefix}{Output name prefix for the .Rdata file containing window counts, background counts
and filtering statistics calculated during TSS detection.}

\item{foldChange}{A fold change cutoff of local enrichment to detect the TSS. For samples with
usual' amount of starting material and squencing depth (>=5ug starting material,
= 5 mil reads/sample), a cut-off of 6 fold can be used. For samples with low
amount of material or sequencing depth, use a lower cut-off (eg. use 2-fold for
samples with 500ng starting material).}

\item{restrictChr}{Chromosomes to restrict the analysis to.}
}
\value{
.bed files containing TSS position for each group, along with a bed file for consensus
       (union) TSS sites of all samples.
}
\description{
Detection of Trancription start sites based on local enrichment
}
\examples{

# before running this
# 1. Create a CapSet object
# 2. de-multiplex the fastqs
# 3. map them
# 4. filter duplicate reads from mapped BAM

# load a previously saved CapSet object
cs <- exampleCSobject()
# detect TSS (samples in same group are treated as replicates)
cs <- detectTSS(cs, groups = rep(c("wt","mut"), each = 2), outfile_prefix = "testTSS",
           foldChange = 6, restrictChr = "X")

}
